server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java

```diff
@@ -1400,24 +1413,45 @@ public Void apply(@Nullable Object input)
                  .emit();
             }
 
-            droppingSinks.remove(identifier);
-            sinkTimeline.remove(
-                sink.getInterval(),
-                sink.getVersion(),
-                identifier.getShardSpec().createChunk(sink)
+            log.info(
+                "Unannounced segment[%s], scheduling drop in [%d] millisecs",
+                identifier,
+                segmentLoaderConfig.getDropSegmentDelayMillis()
             );
-            for (FireHydrant hydrant : sink) {
-              if (cache != null) {
-                cache.close(SinkQuerySegmentWalker.makeHydrantCacheIdentifier(hydrant));
+
+            Runnable removeRunnable = () -> {
+              droppingSinks.remove(identifier);
+              sinkTimeline.remove(
+                  sink.getInterval(),
+                  sink.getVersion(),
+                  identifier.getShardSpec().createChunk(sink)
+              );
+              for (FireHydrant hydrant : sink) {
+                if (cache != null) {
+                  cache.close(SinkQuerySegmentWalker.makeHydrantCacheIdentifier(hydrant));
+                }
+                hydrant.swapSegment(null);
               }
-              hydrant.swapSegment(null);
-            }
 
-            if (removeOnDiskData) {
-              removeDirectory(computePersistDir(identifier));
-            }
+              if (removeOnDiskData) {
+                removeDirectory(computePersistDir(identifier));
+              }
+
+              log.info("Dropped segment[%s].", identifier);
+            };
 
-            log.info("Dropped segment[%s].", identifier);
+            if (segmentLoaderConfig == null) {
+              removeRunnable.run();
+            } else {
+              // Keep the segments in the cache and sinkTimeline for dropSegmentDelay after unannouncing the segments
+              // This way, in transit queries which still see the segments in this peon would be able to query the
+              // segments and not throw NullPtr exceptions.
+              exec.schedule(
```

Comment:
when do we shutdown this exec?

Reply:
Good question. I am mostly follow the pattern in the historical side. Or this [SegmentLoadDropHandler](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java). There, the `exec` is created without any shutdown. 

That said, it may not be a good example there. Any suggestions here?

Reply:
How about following the pattern with `persistExecutor`. Basically, in [shutdownExecutors](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java#L1160) method, I will also call `exec.shutdownNow()`?

Reply:
made the changes